File: erlang/test_async.erl
-module(test_async).
-export([main/0]).

%% ---------------------------------------------------------------------------
%% ASYNCHRONOUS COMMUNICATION
%% From https://go.dev/ref/spec#Channel_types: The capacity, in number of
%% elements, sets the size of the buffer in the channel. If the capacity is
%% zero or absent, the channel is unbuffered and communication succeeds only
%% when both a sender and receiver are ready. Otherwise, the channel is buffered
%% and communication succeeds without blocking if the buffer is not full (sends)
%% or not empty (receives). A nil channel is never ready for communication.
%% ---------------------------------------------------------------------------

test_async_sender(Ch, Msgs) ->
    test_async_sender(Ch, Msgs, 0).

test_async_sender(Ch, Msgs, I) when I < Msgs ->
    goencoding:async_send(Ch, I), % does not block if buffer has space
    io:format("[async] Sender sent: ~p~n", [I]),
    test_async_sender(Ch, Msgs, I + 1);

test_async_sender(Ch, _Msgs, _I) ->
    goencoding:async_close(Ch). % signal "no more values"

test_async_receiver(Ch) ->
    case goencoding:async_recv(Ch) of % loops until channel is closed
        {ok, Msg} ->
            timer:sleep(100), % simulate slow processing
            io:format("[async] Receiver got: ~p~n", [Msg]),
            test_async_receiver(Ch);
        closed ->
            io:format("[async] Receiver: channel closed, no more messages, done receiving~n")
    end.

test_async() ->
    Ch = goencoding:async_new(2), % buffered channel with capacity 2
    Parent = self(),

    io:format("[async] Spawned receiver goroutine...~n"),
    spawn(fun() ->
        test_async_receiver(Ch),
        Parent ! {done, receiver}
    end),

    timer:sleep(100), % Ensure receiver is ready before sender starts

    io:format("[async] Spawned sender goroutine...~n"),
    spawn(fun() ->
        test_async_sender(Ch, 5), % send 5 messages (buffer is only 2)
        Parent ! {done, sender}
    end),

    % Wait for both to complete
    receive {done, receiver} -> ok end,
    receive {done, sender} -> ok end.


test_async_multiple_blocked_senders() ->
    Ch = goencoding:async_new(2), % buffered channel with capacity 2
    Parent = self(),

    % Fill the buffer to block senders
    goencoding:async_send(Ch, 1),
    goencoding:async_send(Ch, 2),
    io:format("[async: multi-sender] Buffer filled with [1, 2]~n"),

    io:format("[async] Spawned receiver goroutine...~n"),
    spawn(fun() ->
        test_async_receiver(Ch),
        Parent ! {done, receiver}
    end),

    timer:sleep(100), % Ensure receiver is ready before sender starts

    io:format("[async] Spawned sender goroutine...~n"),
    spawn(fun() ->
        goencoding:async_send(Ch, 3),
        goencoding:async_send(Ch, 4),
        goencoding:async_send(Ch, 5),
        goencoding:async_close(Ch), % signal "no more values"
        Parent ! {done, sender}
    end),

    % Wait for both to complete
    receive {done, receiver} -> ok end,
    receive {done, sender} -> ok end.


%% ---------------------------------------------------------------------------
%% MAIN
%% ---------------------------------------------------------------------------

main() ->
    io:format("~n--- Asynchronous tests ---~n"),

    io:format("------ Async 1: Send on an open channel (one sender, one receiver) ------~n"),
    test_async(),
	
    io:format("~n------ Async 2: Send on an open channel (multiple senders, one receiver) ------~n"),
    test_async_multiple_blocked_senders(),

	io:format("--- End Asynchronous tests ---~n").

File: go/test_async.go
package main

import (
	"fmt"
	"sync"
	"time"
)

// ---------------------------------------------------------------------------
// ASYNCHRONOUS COMMUNICATION
// From https://go.dev/ref/spec#Channel_types: The capacity, in number of
// elements, sets the size of the buffer in the channel. If the capacity is
// zero or absent, the channel is unbuffered and communication succeeds only
// when both a sender and receiver are ready. Otherwise, the channel is buffered
// and communication succeeds without blocking if the buffer is not full (sends)
// or not empty (receives). A nil channel is never ready for communication.
// ---------------------------------------------------------------------------
func test_async_sender(ch chan int, msgs int) {
	for i := 0; i < msgs; i++ {
		ch <- i // does not block if buffer has space
		fmt.Printf("[async: single sender] Sender sent: %d\n", i)
	}
	close(ch) // signal "no more values"
}

func test_async_receiver(ch chan int) {
	for msg := range ch { // loops until channel is closed
		fmt.Println("[async: single sender] Receiver got:", msg)
		time.Sleep(200 * time.Millisecond) // Simulate slower receiver to demonstrate buffering
	}
	fmt.Println("[async: single sender] Receiver: channel closed, no more messages, done receiving")
}

func test_async() {
	ch := make(chan int, 2) // buffered channel with capacity 2
	var wg sync.WaitGroup

	wg.Add(2)

	fmt.Println("[async: single sender] Spawned receiver goroutine...")
	go func() {
		defer wg.Done()
		test_async_receiver(ch)
	}()

	time.Sleep(100 * time.Millisecond) // Ensure receiver is ready before sender starts

	fmt.Println("[async: single sender] Spawned sender goroutine...")
	go func() {
		defer wg.Done()
		test_async_sender(ch, 5) // send 5 messages (buffer is only 2)
	}()

	wg.Wait() // blocks until all Done() are called
}

func test_async_multiple_blocked_senders() {
	ch := make(chan int, 2) // buffered channel with capacity 2
	var wg sync.WaitGroup

	// Fill the buffer first
	ch <- 1
	ch <- 2
	fmt.Println("[async: multi-sender] Buffer filled with [1, 2]")

	// Spawn 3 senders that will all block (buffer is full)
	for i := 3; i <= 5; i++ {
		wg.Add(1)
		senderID := i
		go func() {
			defer wg.Done()
			fmt.Printf("[async: multi-sender] Sender %d attempting to send...\n", senderID)
			ch <- senderID // This will block because buffer is full
			fmt.Printf("[async: multi-sender] Sender %d completed send\n", senderID)
		}()
	}

	time.Sleep(100 * time.Millisecond) // Let all senders block

	fmt.Println("[async: multi-sender] All 3 senders are now blocked. Starting receiver...")

	// Now receive messages one by one
	// Each receive should unblock one sender
	for i := 1; i <= 5; i++ {
		msg := <-ch
		fmt.Printf("[async: multi-sender] Receiver got: %d\n", msg)
		time.Sleep(200 * time.Millisecond) // Slow down to see sender unblocking
	}

	wg.Wait()
	close(ch)
	fmt.Println("[async: multi-sender] Test complete")
}

// ---------------------------------------------------------------------------
// MAIN
// ---------------------------------------------------------------------------

func test_async_main() {
	fmt.Println("--- Asynchronous ---")
	test_async()
	test_async_multiple_blocked_senders()
	fmt.Println("--- End Asynchronous --\n")
}
